### DISCUSSION TITLE
How to access Request object & dependencies of FastAPI in models created from Pydantic's BaseModel

### DISCUSSION THREAD

**Discussion by jeetpatel9** [fastapi/fastapi#5959] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@profile_router.post("/forgot-password", status_code=status.HTTP_200_OK)
async def forgot_password(
    payload: ForgotPasswordRequestSchema,
) -> ForgotPasswordResponseSchema:
    """
    This API will send reset password link on mail.
    """

    user payload._user

    return ForgotPasswordResponseSchema()

class ForgotPasswordRequestSchema(BaseModel):
    email: EmailStr
    _session = get_db() # this will simply return the session of database.
    _user = None
    
    @validator("email")
    def validate_email(cls, v):
        # Here I want to query on Users's model to see if the email exist in the 
        # database. If the email does. not exist then I would like to raise a custom 
        # python exception as shown below.

        user = cls._session.get(Users, email=v) # Here I can use session as I have 
        # already initialised it as a class variable.

        if not user:
            cls.session.close()
            raise ValueError("Email does not exist in the database.")

        cls._user = user # this is because we want to use user object in the request 
        # function.

        cls.session.close()

        return v
```

### Description

I am writing APIs using stack FastAPI, Pydantic & SQL Alchemy and I have come across many cases where I had to query database to perform validations on payload values. Let's consider one example API, `/forgot-password`. This API will accept email in the payload and I need to validate the existence of the email in database. If the email exist in the database then necessary action like creating token and sending mail would be performed or else an error response against that field should be raise by Pydantic. The error responses must be the standard `PydanticValueError` response. This is because all the validation errors would have consistent responses as it becomes easy to handle for the consumers.

# Payload -

```
{
    "email": "example@gmail.com"
}
```

```
class ForgotPasswordRequestSchema(BaseModel):
    email: EmailStr
    
    @validator("email")
    def validate_email(cls, v):
        # this is the db query I want to perform but 
        # I do not have access to the active session of this request.
        user = session.get(Users, email=v) 
        if not user:
            raise ValueError("Email does not exist in the database.")

        return v
```

In Pydantic this schema and the validation for email is implemented as -

```
class ForgotPasswordRequestSchema(BaseModel):
    email: EmailStr
    
    @validator("email")
    def validate_email(cls, v):
        # this is the db query I want to perform but 
        # I do not have access to the active session of this request.
        user = session.get(Users, email=v) 
        if not user:
            raise ValueError("Email does not exist in the database.")

        return v
```

Now this can be easily handled if the we simple create an Alchemy session in the pydantic model like this.

```
class ForgotPasswordRequestSchema(BaseModel):
    email: EmailStr
    _session = get_db() # this will simply return the session of database.
    _user = None
    
    @validator("email")
    def validate_email(cls, v):
        # Here I want to query on Users's model to see if the email exist in the 
        # database. If the email does. not exist then I would like to raise a custom 
        # python exception as shown below.

        user = cls._session.get(Users, email=v) # Here I can use session as I have 
        # already initialised it as a class variable.

        if not user:
            cls.session.close()
            raise ValueError("Email does not exist in the database.")

        cls._user = user # this is because we want to use user object in the request 
        # function.

        cls.session.close()

        return v
```

But it is not a right approach as through out the request only one session should be used. As you can see in above example we are closing the session so we won't be able to use the user object in request function as `user = payload._user`. This means we will have to again query for the same row in request function. If we do not close the session then we are seeing alchemy exceptions like this - `sqlalchemy.exc.PendingRollbackError`

Now, the best approach is to be able to use the same session in the Pydantic model which is created at the start of request and is also closing at the end of the request.

So, I am basically looking for a way to pass that session to Pydantic as context. Session to my request function is provided as dependency.

### Operating System

macOS

### FastAPI Version

0.89.1

### Python Version

3.10

**Comment by b4sen** [comment]
Why don't you just inject the DB into the endpoint and do something like this:
```python
@app.post("/forgot-password")
def forgot_password(schema_in: ForgotPasswordRequestSchema, db: Session = Depends(get_db)):
    db_obj  = db.query(User).where(User.email == schema_in.email).one_or_none()
    if not db_obj:
        raise HTTPException(status_code=404, detail="Email not found")
    # continue with the password reset flow
```

you can use pydantic to validate the email string and length of password but I don't think you should use it to query database in the validator

**Reply by jeetpatel9** [reply]
This is good but won't have the standard response for errors.

**Comment by jgould22** [comment]
I would also recommend against preforming db queries or external calls in the pydantic validators.

I had done it the way you described and it became painful to work around when I needed to use async code in the validator.

If you dont use async code to call out in the validator you could block the event loop during validation on async endpoints.

I am interested to see if there is a better way.

**Comment by rochacbruno** [comment]
I am solving like this:

```python
@router.post("/pwd_reset_token/")
async def send_password_reset_token(
    *,
    email: EmailStr = Body(embed=True),
    background_tasks: BackgroundTasks,
):
    background_tasks.add_task(try_to_send_pwd_reset_email, email=email)
    return {"message": "If we found a user with that email, we sent a password reset token to it."}

```

Then inside the task I perform the database queries to validate the user existence.

`tasks/user.py
```python
def try_to_send_pwd_reset_email(email):
    """Given an email address sends email if user is found"""
    with Session(engine) as session:
        user = session.exec(select(User).where(User.email == email)).first()
        if not user:
            return

        sender = settings.email.smtp_sender  # pyright: ignore
        url = settings.security.PWD_RESET_URL  # pyright: ignore
        expire = settings.security.RESET_TOKEN_EXPIRE_MINUTES  # pyright: ignore

        pwd_reset_token = create_access_token(
            data={"sub": user.username},
            expires_delta=timedelta(minutes=expire),  # pyright: ignore
            scope="pwd_reset",
        )

        send_email(
            email=user.email,
            message=MESSAGE.format(
                sender=sender,
                to=user.email,
                url=url,
                pwd_reset_token=pwd_reset_token,
                expire=expire,
            ),
        )
```

[repo](https://github.com/rochacbruno/dundie-api/tree/end_result/dundie)

**Comment by anubhav-sl** [comment, marked as answer]
**It is not recommended to query the database in pydantic schema.** Instead [use session as a dependency][1].

If you want to raise errors like pydantic validation error you might need this:

```
def raise_custom_error(exc: Exception, loc: str, model: BaseModel, status_code=int, **kwargs):
    """
    This method will return error responses using pydantic error wrapper (similar to pydantic validation error).
    """
    raise HTTPException(
        detail=json.loads(ValidationError([ErrorWrapper(exc(**kwargs), loc=loc)], model=model).json()),
        status_code=status_code,
    )

```
<h3>
Usage
</h3>

```
class PayloadSchema(BaseModel):
    email: EmailStr

@app_router.post('/forgot-password')
def forgot_password(
    payload: PayloadSchema,
    session: Session = Depends(get_db),
    background_tasks: BackgroundTasks
):
    
    existing_user = db.get(Users, email=payload.email)
    if(existing_user):
        raise_custom_error(
        PydanticValueError, "email", PayloadSchema, status.HTTP_400_BAD_REQUEST
    )
    background_tasks(send_email, email=payload.email)

```


  [1]: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#a-database-dependency-with-yield

**Reply by jeetpatel9** [reply]
This seems like a better approach. I will be able to maintain the return standard response for all the validation errors.

Thanks @anubhav-sl
